src/tests/suite/daos_checksum.c

```diff
@@ -679,24 +702,204 @@ ARRAY_UPDATE_FETCH_TESTCASE(state, {
+}
+
+static void
+test_update_fetch_a_key(void **state)
+{
+	key_csum_fetch_update(state,
+			      DAOS_CHECKSUM_UPDATE_AKEY_FAIL,
+			      DAOS_CHECKSUM_FETCH_AKEY_FAIL);
+}
+
+static void
+test_update_fetch_d_key(void **state)
+{
+	key_csum_fetch_update(state,
+			      DAOS_CHECKSUM_UPDATE_DKEY_FAIL,
+			      DAOS_CHECKSUM_FETCH_DKEY_FAIL);
+
+}
+
+#define KDS_NR 10
+static void
+test_enumerate_a_key(void **state)
+{
+	struct csum_test_ctx	ctx = {0};
+	int			rc;
+	uint32_t		i;
+	daos_anchor_t		anchor = {0};
+	daos_key_desc_t		kds[KDS_NR] = {0};
+	d_sg_list_t		sgl = {0};
+	uint32_t		nr = KDS_NR;
+
+	setup_from_test_args(&ctx, *state);
+	setup_cont_obj(&ctx, DAOS_PROP_CO_CSUM_CRC16, false, 1024, OC_SX);
+	setup_simple_data(&ctx);
+
+	/** insert multiple keys to enumerate */
+	for (i = 0; i < KDS_NR; i++) {
+		rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, 0, &ctx.dkey, 1,
+				     &ctx.update_iod, &ctx.update_sgl,
+				     NULL);
+		assert_int_equal(0, rc);
+		((uint8_t *)ctx.update_iod.iod_name.iov_buf)[0] += 1;
+	}
+
+	/** Make sure can handle verifying keys over multiple iovs */
+	d_sgl_init(&sgl, 2);
+	iov_alloc(&sgl.sg_iovs[0], 10);
+	iov_alloc(&sgl.sg_iovs[1], 100);
+
+	/** inject failure ... should return CSUM error */
+	set_fetch_akey_fi();
+	rc = daos_obj_list_akey(ctx.oh, DAOS_TX_NONE, &ctx.dkey, &nr, kds, &sgl,
+			   &anchor, NULL);
+	assert_int_equal(-DER_CSUM, rc);
+	unset_csum_fi();
+
+	/** Sanity check that no failure still returns success */
+	nr = 10;
+	memset(&anchor, 0, sizeof(anchor));
+	rc = daos_obj_list_akey(ctx.oh, DAOS_TX_NONE, &ctx.dkey, &nr, kds, &sgl,
+				&anchor, NULL);
+	assert_int_equal(0, rc);
+	assert_int_equal(KDS_NR, nr);
+
+	/** Clean up */
+	d_sgl_fini(&sgl, true);
+	cleanup_data(&ctx);
+	cleanup_cont_obj(&ctx);
+}
+
+static void
+test_enumerate_d_key(void **state)
+{
+	struct csum_test_ctx	ctx = {0};
+	int			rc;
+	uint32_t		i;
+	daos_anchor_t		anchor = {0};
+	daos_key_desc_t		kds[KDS_NR] = {0};
+	d_sg_list_t		sgl = {0};
+	uint32_t		nr = KDS_NR;
+
+	setup_from_test_args(&ctx, *state);
+	setup_cont_obj(&ctx, DAOS_PROP_CO_CSUM_CRC16, false, 1024, OC_SX);
+	setup_simple_data(&ctx);
+
+	/** insert multiple keys to enumerate */
+	for (i = 0; i < KDS_NR; i++) {
+		rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, 0, &ctx.dkey, 1,
+				     &ctx.update_iod, &ctx.update_sgl,
+				     NULL);
+		assert_int_equal(0, rc);
+		((uint8_t *)ctx.dkey.iov_buf)[0] += 1;
+	}
+
+	/** Make sure can handle verifying keys over multiple iovs */
+	d_sgl_init(&sgl, 2);
+	iov_alloc(&sgl.sg_iovs[0], 10);
+	iov_alloc(&sgl.sg_iovs[1], 100);
+
+	/** inject failure ... should return CSUM error */
+	set_fetch_dkey_fi();
+	rc = daos_obj_list_dkey(ctx.oh, DAOS_TX_NONE, &nr, kds, &sgl,
+			   &anchor, NULL);
+	assert_int_equal(-DER_CSUM, rc);
+	unset_csum_fi();
+
+	/** Sanity check that no failure still returns success */
+	nr = KDS_NR;
+	memset(&anchor, 0, sizeof(anchor));
+	uint32_t key_count = 0;
+	while(!daos_anchor_is_eof(&anchor)) {
```

Comment:
(style)  Missing a blank line after declarations
(style)  space required before the open parenthesis '('